core/src/main/java/org/bitcoinj/core/Peer.java

```diff
@@ -451,6 +459,12 @@ protected void processMessage(Message m) throws Exception {
             currentFilteredBlock = null;
         }
 
+        // No further communication is possible until version handshake is complete.
+        if (!(m instanceof VersionMessage || m instanceof VersionAck
+                || (versionHandshakeFuture.isDone() && !versionHandshakeFuture.isCancelled())))
```

Comment:
Interesting, if you call isDone() or isCancelled() on a list of futures, it's equivalent to checking all of the futures and ANDing their responses?